jsonrpc/server.go

```diff
@@ -211,59 +213,76 @@ func (s *Server) HandleReader(ctx context.Context, reader io.Reader) ([]byte, er
 		}
 	} else {
 		var batchReq []json.RawMessage
-		var batchRes []json.RawMessage
 
 		if batchJSONErr := dec.Decode(&batchReq); batchJSONErr != nil {
 			res.Error = Err(InvalidJSON, batchJSONErr.Error())
 		} else if len(batchReq) == 0 {
 			res.Error = Err(InvalidRequest, "empty batch")
 		} else {
-			for _, rawReq := range batchReq { // todo: handle async
-				var resObject *response
+			return s.handleBatchRequest(ctx, batchReq)
+		}
+	}
+
+	if res == nil {
+		return nil, nil
+	}
+	return json.Marshal(res)
+}
+
+func (s *Server) handleBatchRequest(ctx context.Context, batchReq []json.RawMessage) ([]byte, error) {
+	var batchRes []json.RawMessage
+	var resMutex sync.Mutex
 
-				reqDec := json.NewDecoder(bytes.NewBuffer(rawReq))
-				reqDec.UseNumber()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-				req := new(request)
-				if jsonErr := reqDec.Decode(req); jsonErr != nil {
+	p := pool.New()
```

Comment:
We should limit concurrency here else it could be a DoS vector.

Can we make pool a shared thing in `Server`? So that all requests use the same pool for concurrency and we dont overwhelm the scheduler at the times of heavy RPC usage.

Reply:
Good point, do you think server should accept size of pool in constructor or we can hardcode its value to some reasonable value?

Reply:
`runtime.GOMAXPROCS(0)` should be good enough for now.

Reply:
Done. It requires additional `WaitGroup` var  (`pool.Wait()` might wait for parallel batch request)

Reply:
> Done. It requires additional WaitGroup var

Yeap, that is fine.